pkg/awsiamauth/reconciler/reconciler.go

```diff
@@ -0,0 +1,123 @@
+package reconciler
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/google/uuid"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	anywherev1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/awsiamauth"
+	anywhereCluster "github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/constants"
+	"github.com/aws/eks-anywhere/pkg/controller"
+	"github.com/aws/eks-anywhere/pkg/controller/clientutil"
+	"github.com/aws/eks-anywhere/pkg/controller/serverside"
+	"github.com/aws/eks-anywhere/pkg/crypto"
+)
+
+// AWSIamConfigReconciler is an interface for reconciling AWSIAMConfig in the Cluster reconciler.
+type AWSIamConfigReconciler interface {
+	ReconcileAWSIAMAuthCASecret(ctx context.Context, logger logr.Logger, client client.Client, clusterName string) (controller.Result, error)
+	Reconcile(ctx context.Context, logger logr.Logger, client client.Client, cluster *anywherev1.Cluster) (controller.Result, error)
+}
+
+// RemoteClientRegistry is an interface that defines methods for remote clients.
+type RemoteClientRegistry interface {
+	GetClient(ctx context.Context, cluster client.ObjectKey) (client.Client, error)
+}
+
+// Reconciler allows to reconcile AWSIamConfig.
+type Reconciler struct {
+	certgen              crypto.CertificateGenerator
+	templateBuilder      *awsiamauth.TemplateBuilder
+	clusterID            uuid.UUID
+	remoteClientRegistry RemoteClientRegistry
+}
+
+// New returns a new Reconciler.
+func New(certgen crypto.CertificateGenerator, clusterID uuid.UUID, remoteClientRegistry RemoteClientRegistry) *Reconciler {
+	return &Reconciler{
+		certgen:              certgen,
+		templateBuilder:      &awsiamauth.TemplateBuilder{},
+		clusterID:            clusterID,
+		remoteClientRegistry: remoteClientRegistry,
+	}
+}
+
+// ReconcileAWSIAMAuthCASecret ensures the AWS IAM Authenticator secret is present.
+// It uses a controller.Result to indicate when requeues are needed.
+func (r *Reconciler) ReconcileAWSIAMAuthCASecret(ctx context.Context, logger logr.Logger, client client.Client, clusterName string) (controller.Result, error) {
```

Comment:
Why don't we have the client injected as a dependency in the constructor?

Reply:
Yup that seems better 👍 Initially I was not sure if I would need client or remoteClient more so i was just passing it in.